### ISSUE TITLE
Audio: How does Emscripten calculate its spec?

### ISSUE THREAD

**Issue by km19809** [libsdl-org/SDL#6060]
I saw a recent commit (https://github.com/libsdl-org/SDL/commit/90b86b132af02d0a8b1d39103b04ebb667687fec).
It should handle the non-power-of-two samples problem.
However, I initialized SDL_audio, which compiled with emscripten, it showed an error like this:
![image](https://user-images.githubusercontent.com/20869932/184618957-9c3571ca-9fd9-4624-a9e3-4ca821f261f6.png)

(Sorry, I could not prepare a minimal reproducible example yet)

I looked into [emscriptenaudio.c](https://github.com/libsdl-org/SDL/blob/main/src/audio/emscripten/SDL_emscriptenaudio.c#L331-L340), and I could find that the `spec.samples` was 1470.
However, I could not find how the spec was calculated, even after I examined compiled js codes.

How does emscripten calculate its spec?

Also, if I modify `EMSCRIPTENAUDIO_Init` to set `Impl->SupportsNonPow2Samples = SDL_FALSE`, will it fix this issue?

**Comment by icculus** [comment]
> Also, if I modify EMSCRIPTENAUDIO_Init to set Impl->SupportsNonPow2Samples = SDL_FALSE, will it fix this issue?

It's already SDL_FALSE, since `Impl` is zeroed before that call.

Are you sure you're using an SDL with that change? It shouldn't have been able to reach `EMSCRIPTENAUDIO_OpenDevice` with a non-power-of-two sample size. That change will be in 2.24.0 (which is supposed to be released this week), but it hasn't been in an officially released version yet.

**Comment by km19809** [comment]
It's my mistake. I used an SDL2 version that emscripten automatically fetched.
I did not know the commit is not released yet.
However, I personally modified the compiled js to give power of two, but it still throws an error.
`Web Audio capture buffer length mismatch! Destination size: 2048 samples vs expected 1470 samples!`
The message comes from [here](https://github.com/libsdl-org/SDL/blob/main/src/audio/emscripten/SDL_emscriptenaudio.c#L115), so I think `(this->spec.size / sizeof (float)) / this->spec.channels` is 1470.
But I cannot find where those `spec` coming from, so I cannot debug it too.
Can you give me some hints?

**Comment by icculus** [comment]
I'm still confused as to what is going wrong:

First: where did 1470 come from? When we [create the audio node](https://github.com/libsdl-org/SDL/blob/dc62fec5e99add18d2805743a3dda53826003c2b/src/audio/emscripten/SDL_emscriptenaudio.c#L297) ...

```c
SDL2.capture.scriptProcessorNode = SDL2.audioContext.createScriptProcessor($1, $0, 1);
```

The `$1` is the number of samples we request, and [this should fail if not a power of two](https://developer.mozilla.org/en-US/docs/Web/API/BaseAudioContext/createScriptProcessor#parameters).

So I guess the questions are:
- Where did 1470 come from? Did the app calling SDL_OpenAudio (or SDL_OpenAudioDevice) specify that value when opening the device?
- Is `_this->spec.samples` set to 1470 when `EMSCRIPTENAUDIO_OpenDevice` is called?

The good news, if this for some reason is giving us a smaller buffer than we expected, is that SDL's audio capture code is designed to deal with this scenario, and we can just remove the error and return whatever data Emscripten is making available at the moment, but this feels like a bug we can fix before we get to that point.

**Comment by icculus** [comment]
Just tried this with the latest SDL (and emsdk 2.0.34), and it's working for me on Firefox. I tried our testaudiocapture program with these changes:

```diff
diff --git a/test/testaudiocapture.c b/test/testaudiocapture.c
index 03d373273..b042fdd0a 100644
--- a/test/testaudiocapture.c
+++ b/test/testaudiocapture.c
@@ -121,7 +121,7 @@ main(int argc, char **argv)
     wanted.freq = 44100;
     wanted.format = AUDIO_F32SYS;
     wanted.channels = 1;
-    wanted.samples = 4096;
+    wanted.samples = 1470;
     wanted.callback = NULL;
 
     SDL_zero(spec);
@@ -133,7 +133,9 @@ main(int argc, char **argv)
        circumstances. */
 
     SDL_Log("Opening default playback device...\n");
+    SDL_Log("Opening playback device with %d samples", (int) wanted.samples);
     devid_out = SDL_OpenAudioDevice(NULL, SDL_FALSE, &wanted, &spec, SDL_AUDIO_ALLOW_ANY_CHANGE);
+    SDL_Log("Playback device has %d samples", (int) spec.samples);
     if (!devid_out) {
         SDL_LogError(SDL_LOG_CATEGORY_APPLICATION, "Couldn't open an audio device for playback: %s!\n", SDL_GetError());
         SDL_Quit();
@@ -145,7 +147,10 @@ main(int argc, char **argv)
             devname ? devname : "[[default]]",
             devname ? "'" : "");
 
+spec.samples = 1470;
+    SDL_Log("Opening capture device with %d samples", (int) spec.samples);
     devid_in = SDL_OpenAudioDevice(argv[1], SDL_TRUE, &spec, &spec, 0);
+    SDL_Log("Opened capture device has %d samples", (int) spec.samples);
     if (!devid_in) {
         SDL_LogError(SDL_LOG_CATEGORY_APPLICATION, "Couldn't open an audio device for capture: %s!\n", SDL_GetError());
         SDL_Quit();
```

And it worked--recording and playback--and reported this:

```
INFO: Opening default playback device...
INFO: Opening playback device with 1470 samples
INFO: Playback device has 2048 samples
INFO: Opening capture device [[default]]...
INFO: Opening capture device with 1470 samples
INFO: Opened capture device has 1470 samples
```

The playback device was opened with `SDL_AUDIO_ALLOW_ANY_CHANGE`, and SDL correctly bumped it to a power-of-two sample count. The capture device was _not_, so SDL is actually buffering between the 1470 samples the app expects and the 2048 that the web browser expects (had I allowed the change, this would remove the buffer and just tell the app to expect 2048 samples).

So I'm guessing there was some problem with either the SDL build being used or the hand-editing that was done for testing in https://github.com/libsdl-org/SDL/issues/6060#issuecomment-1215019541 ... so I _think_ that when 2.24.0 ships in the next few days (and Emscripten picks up the new version), this problem is going to fix itself.

If it doesn't, please let me know and we'll reopen the issue!

**Comment by km19809** [comment]
> Where did 1470 come from? Did the app calling SDL_OpenAudio (or SDL_OpenAudioDevice) specify that value when opening the device?
Is _this->spec.samples set to 1470 when EMSCRIPTENAUDIO_OpenDevice is called?

Yes, these are my question.
I was not able to find out how these values are calculated.
It's because the compiled wasm code was hard to follow.
So I decided to ask questions here.

I think I have to start with building a minimal reproducible example.
The current code is an experiment that integrates SDL2, and Rust with Emscripten, so it's complex.

**Comment by km19809** [comment]
> Just tried this with the latest SDL (and emsdk 2.0.34), and it's working for me on Firefox. I tried our testaudiocapture program with these changes:
> 
> ```diff
> diff --git a/test/testaudiocapture.c b/test/testaudiocapture.c
> index 03d373273..b042fdd0a 100644
> --- a/test/testaudiocapture.c
> +++ b/test/testaudiocapture.c
> @@ -121,7 +121,7 @@ main(int argc, char **argv)
>      wanted.freq = 44100;
>      wanted.format = AUDIO_F32SYS;
>      wanted.channels = 1;
> -    wanted.samples = 4096;
> +    wanted.samples = 1470;
>      wanted.callback = NULL;
>  
>      SDL_zero(spec);
> @@ -133,7 +133,9 @@ main(int argc, char **argv)
>         circumstances. */
>  
>      SDL_Log("Opening default playback device...\n");
> +    SDL_Log("Opening playback device with %d samples", (int) wanted.samples);
>      devid_out = SDL_OpenAudioDevice(NULL, SDL_FALSE, &wanted, &spec, SDL_AUDIO_ALLOW_ANY_CHANGE);
> +    SDL_Log("Playback device has %d samples", (int) spec.samples);
>      if (!devid_out) {
>          SDL_LogError(SDL_LOG_CATEGORY_APPLICATION, "Couldn't open an audio device for playback: %s!\n", SDL_GetError());
>          SDL_Quit();
> @@ -145,7 +147,10 @@ main(int argc, char **argv)
>              devname ? devname : "[[default]]",
>              devname ? "'" : "");
>  
> +spec.samples = 1470;
> +    SDL_Log("Opening capture device with %d samples", (int) spec.samples);
>      devid_in = SDL_OpenAudioDevice(argv[1], SDL_TRUE, &spec, &spec, 0);
> +    SDL_Log("Opened capture device has %d samples", (int) spec.samples);
>      if (!devid_in) {
>          SDL_LogError(SDL_LOG_CATEGORY_APPLICATION, "Couldn't open an audio device for capture: %s!\n", SDL_GetError());
>          SDL_Quit();
> ```
> 
> And it worked--recording and playback--and reported this:
> 
> ```
> INFO: Opening default playback device...
> INFO: Opening playback device with 1470 samples
> INFO: Playback device has 2048 samples
> INFO: Opening capture device [[default]]...
> INFO: Opening capture device with 1470 samples
> INFO: Opened capture device has 1470 samples
> ```
> 
> The playback device was opened with `SDL_AUDIO_ALLOW_ANY_CHANGE`, and SDL correctly bumped it to a power-of-two sample count. The capture device was _not_, so SDL is actually buffering between the 1470 samples the app expects and the 2048 that the web browser expects (had I allowed the change, this would remove the buffer and just tell the app to expect 2048 samples).
> 
> So I'm guessing there was some problem with either the SDL build being used or the hand-editing that was done for testing in [#6060 (comment)](https://github.com/libsdl-org/SDL/issues/6060#issuecomment-1215019541) ... so I _think_ that when 2.24.0 ships in the next few days (and Emscripten picks up the new version), this problem is going to fix itself.
> 
> If it doesn't, please let me know and we'll reopen the issue!

Oh, I checked this comment now.
After release, I'll try it.
Thank you!
